### ISSUE TITLE
coreclr.dll versioning and the .version file. 

### ISSUE THREAD

**Issue by gregkalapos** [dotnet/coreclr#11225]
Tldr:
What is the best and easiest in-process way to figure out the version of the CoreCLR on which an application runs? (e.g. for a monitoring tool) 

More details on this:
Pre 1.1 the coreclr.dll version somewhat revealed the CoreCLR version (e.g. for 1.0.1 it’s 1.0.24410.1, for 1.03 it’s 1.0.24709.1, etc.). But this was not consistence in the 1.1 Release, since in 1.1.0 it’s 4.6.24628.01, for 1.1.1 it’s 4.6.25009.03.

There is also this ‘.version’ file, which seem to have the exact version I want to automatically report, but I don’t find any documentation on it. Is it always two lines, one with a hash and then the version? Or could that look differently? Plus if I create a self-contained application (by adding e.g. <RuntimeIdentifier>win10-x64</RuntimeIdentifier> in the .csproj file) then there is no .version file in the publish folder.  

So in sum, I would like to have answer to these 3 questions:

- Is there a general way to determine the coreclr version in-process for any scenario (not only asp.net core)?
- Is there any thinking behind the versioning of coreclr.dll, or is it just random?
- What is the purpose and the format of the .version file, and is it mandatory?  

Many thanks!

**Comment by gkhanna79** [comment]
@dagood Can help with this question.

**Comment by dagood** [comment]
There's a proposal to be able to do this easily here: https://github.com/dotnet/corefx/issues/16629. Until then, I don't think there's a way that works in all cases.

A workaround @eerhardt suggested last time this came up is to check the folder name of the running shared framework like CLI does, but that doesn't work for standalone apps: https://github.com/dotnet/cli/blob/51dc988fa21cd28ea3e4077bd1f885260b8d99f9/src/Microsoft.DotNet.Cli.Utils/Muxer.cs#L17-L24. Correct me if there's new info here.

A tidbit on the `.version` file: it's created in [Core-Setup](https://github.com/dotnet/core-setup) as part of the shared framework, and the hash it contains is a Core-Setup commit, not a CoreCLR commit. I don't know if the format's stable.

---

About the versioning on coreclr.dll itself, a good place to start if you want to drill down into the code is [versioning.targets](https://github.com/dotnet/buildtools/blob/ff93f91be4c7889442240e1e2b859c335d2191f6/src/Microsoft.DotNet.Build.Tasks/PackageFiles/versioning.targets).

For example, 4.6.24628.01 corresponds to `MajorVersion`, `MinorVersion`, and `BuildNumberMajor/Minor`. 24628.01 was built 246 months after 1996/4/1, on the 28th day, and was the 01st build on that day for that branch. Because minor is only unique to that branch, a build from `master` might have the same version as a build from `release/1.1.0` on the same day--so even if you had a mapping of build number to release version, that won't be 100% reliable if you don't already know which branch it's from.

It looks like the `1.0.*.*` versions were a mistake and 4.6 was always intended: https://github.com/dotnet/coreclr/pull/6578. I don't know precisely what the 4.6 corresponds to, though.

One thing you can get from the DLL itself that might be very useful is the commit hash in the ProductVersion (PowerShell just to show it):

```powershell
(Get-Item coreclr.dll).VersionInfo.ProductVersion
# 4.6.25009.03 built by: dlab-DDVSOWINAGE002. Commit Hash: 1735a1d453677717e68803da6a85284d15dca891
```

**Comment by eerhardt** [comment]
What exactly is it you want?  Just some sort of version that will uniquely tell you if you are on a different runtime?  Does the version need to match/correspond to any of our "marketing" versions (i.e. .NET Core 1.0, 1.1, 2.0)?  Is it explicitly "coreclr" that you are looking for?

We have an API: https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.runtimeinformation.frameworkdescription?view=netstandard-1.6#System_Runtime_InteropServices_RuntimeInformation_FrameworkDescription that may be what you are looking for.  If it isn't, please elaborate on the scenario you are trying to achieve.

**Comment by gregkalapos** [comment]
The proposed ‘RuntimeVersions‘ in #16629 is actually what I’d need. 

> What exactly is it you want? 

So, to describe a scenario: let’s say we implement a monitoring tool, which is used by big organizations with strict policies installed versions, etc. There are people responsible for operation, who may not be C# coders. For a distributed .NET Core system running on 50 machines when a new update comes out we want to report on which machine which runtime is installed, so they have an overview about rolling out new versions. 

> Just some sort of version that will uniquely tell you if you are on a different runtime? Does the version need to match/correspond to any of our "marketing" versions (i.e. .NET Core 1.0, 1.1, 2.0)? Is it explicitly "coreclr" that you are looking for?

Yes and Yes. This is exactly the point here. Imagine where people are responsible for operation, who are not hardcore .NET devs. When .NET Core 2.0 comes out they install it, the devs update the apps then they deploy the updates. When they use our tool we want to tell them that after the update that they use CoreCLR 2.0.0. all machines. If we report 4.6.x.x. they will be confused and we have to explain this to them. 

> To https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.runtimeinformation.frameworkdescription?view=netstandard-1.6#System_Runtime_InteropServices_RuntimeInformation_FrameworkDescription 

 
So this describes the framework, not the runtime (I guess this is addressed in #16629), and it is not in line with the “marketing” versions. Plus, we want to be Netstandard compatible, and (correct me if I’m wrong) this API is not part of the netstandard, and we really don’t want to take up additional dependencies (because that would be added every time to every applications we monitor). Btw. I guess this would also be a problem in #16629 if that is not part of netstandard. 

> A workaround @eerhardt suggested last time this came up is to check the folder name of the running shared framework like CLI does, but that doesn't work for standalone apps.

Thanks for that! Our workaround currently is to look for the .version file and parse that.

> It looks like the 1.0.*.* versions were a mistake and 4.6 was always intended: #6578.

Good to know! 

@eerhardt  @dagood: Thanks for the responses, it helps a lot. I close this now and will watch #6578
